### PR TITLE
New conditional C Scanner; Improvements for virtual C Pre-Processor

### PR DESCRIPTION
**Pull request by ivankravets**
Hi all!

I've finally found a time to make a PR with conditional C Scanner which we use 2 years in production for @PlatformIO. This PR is continue of a work on https://bitbucket.org/scons/scons/pull-requests/380/improvements-for-scons-c-c-pre-processor/diff

I made it maximum compatible with existing API, I even didn't touch original tests. The only incoming data for tests were extended to show that new changes don't break previous behaviour and provide a new conditional C Scanner.

Documentation? This is very complex question for me. I didn't check this item in a list below because I don't know where these changes should be reflected in documentation.

A default classic C Scanner is hardcoded [here](https://github.com/SCons/scons/blob/master/src/engine/SCons/Tool/__init__.py#L84). Should we make additional option for SCons command which will allow to switch between Scanners? 

Nevertheless, new conditional Scanner is a good addition for developers who create an own high level wrapper around SCons. We do this in PlatformIO and use `CConditionalScanner` to build a library dependency tree before calling SCons builder.

I'm ready for any corrections and any comments are welcome.

# Changes

- New conditional C Scanner (`SCons.Scanner.C.CConditionalScanner()`)
  which interprets C/C Preprocessor conditional syntax (#ifdef, #if, #else,
  #elif, #define, etc.)
- Improvements for virtual C Pre-Processor:
  * Handle UNSIGNED LONG and LONG numeric constants in DEC (keep support for HEX)
  * Skip unrecognized directives, such as `#if( defined ...)`
  * Ignore `#include DYNAMIC_INCLUDE` directive that depends on a dynamic
    macro which is not located in a state TABLE.
  * Cleanup CPP expressions before evaluating (strip comments, carriage returns)

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [ ] I have updated the appropriate documentation

### CODE REVIEW

**Comment by bdbaddog** [comment]
Thanks for migrating this pull request over to Github.

File: src/CHANGES.txt [hunk 1]
```diff
@@ -222,22 +226,6 @@ RELEASE 3.0.2 - Mon, 31 Dec 2018 16:00:12 -0700
     - Typo in customized decider example in user guide
     - Replace usage of unittest.TestSuite with unittest.main() (fix #3113)
 
-  From Arda Fu
-    - Fix cpp scanner regex logic to treat ifndef for py3.5+. Previously it was 
-      not properly differentiating between if, ifdef, and ifndef.
-
-
-  From Ivan Kravets, PlatformIO
```
Comment (bdbaddog) [inline src/CHANGES.txt]: Looks like you dropped this in the merge..

**Comment by bdbaddog** [comment]
I'd suggest adding an env variable like CC_SCANNER which defaults to current scanner, but can be assigned to your new one. That way we're not changing default behavior.

For docs you'd just had to add info on that new variable and the possible values.
Maybe add a section to Users guide section 27, miscellaneous functionality.

Also add a wiki page?

Tests would reset the CC_SCANNER to this more complete one and then run same tests as the current c scanner.
That should be sufficient to be able to merge this.

**Comment by bdbaddog** [comment]
I'd love to merge this in as an optional scanner. 
If I'm reading it correctly you've modified the existing scanner.
Given the depth of the change we'd minimally have to go through a deprecation cycle and do some evaluation of the performance vs existing scanner.

If you can change this to be a second optional C/C++ scanner. Then I can merge it.

**Comment by ivankravets** [comment]
Thanks!

1. I've just pushed new changes to our @PlatformIO branch, now you will see only real modified files
2. Yes, you are right, we created a separate scanner named [CConditionalScanner](https://github.com/SCons/scons/pull/3053/files#diff-38f7092e4b890fe35dd54e4e094f319dR128)
3. "Given the depth ". Do you mean C-Preprocessor? The default value for depth is `-1` which is the same which we have now in the master branch of the preprocessor. So, existing projects should not be affected.

**Comment by bdbaddog** [comment]
@ivankravets  - no, by depth I mean how that changing the C/C++ scanner is a nontrivial change.

So if I merge this it will or will not modify the existing C/C++ scanner? 
#2 looks like your answering the inverse of the question I asked.

**Comment by ivankravets** [comment]
I didn't touch original CScanner. You can compare 2 files:
- https://github.com/SCons/scons/blob/master/src/engine/SCons/Scanner/C.py#L111
- https://github.com/platformio/scons/blob/c-conditional-scanner/src/engine/SCons/Scanner/C.py#L138

If someone would like to use a new scanner, he needs `def CConditionalScanner():`.

File: src/engine/SCons/Scanner/C.py [hunk 2]
```diff
@@ -46,16 +46,27 @@ class SConsCPPScanner(SCons.cpp.PreProcessor):
...
     def initialize_result(self, fname):
         self.result = SCons.Util.UniqueList([fname])
-    def finalize_result(self, fname):
-        return self.result[1:]
+
     def find_include_file(self, t):
         keyword, quote, fname = t
-        result = SCons.Node.FS.find_file(fname, self.searchpath[quote])
-        if not result:
+        paths = tuple(self._known_paths) + self.searchpath[quote]
```
Comment (bdbaddog) [inline src/engine/SCons/Scanner/C.py]: you do have modifications to current C scanner here right?

**Comment by ivankravets** [comment]
I've just isolated CConditionalScanner from an original scanner. Indeed, there is a bug in `SConsCPPScanner.find_include_file` which does not resolve correctly includes in `"` quotes which should be found based on the self.current_file directory.

I don't know if it is good if SConsCPPScanner will return the correct found items. Do you mean that it can break existing projects?

**Comment by bdbaddog** [comment]
I'd rather you break that out into a separate pull request.
Is there an existing github issue  for that?
Please separate that out to another PR, then it'll be easier to merge this and then take a look at that issue separately.?

**Comment by ivankravets** [comment]
I already separated, see https://github.com/SCons/scons/blob/fe2680ddd07a4bc7c6064e7193f4979060e178ba/src/engine/SCons/Scanner/C.py#L128

**Comment by bdbaddog** [comment]
You have both changes in this PR right? I'm asking if you can split them into a separate PR for the new scanner and one for the fix to the existing.

**Comment by ivankravets** [comment]
No, I reverted back all my changes to the original scanner and duplicated it with modifications for the conditional preprocessor.

**Comment by ivankravets** [comment]
I added a comment block 

Screen Shot 2020-02-13 at 18 24 44

File: src/engine/SCons/cpp.py [hunk 3]
```diff
@@ -144,12 +154,10 @@
...
 # sequentially, not all at once.
 CPP_to_Python_Eval_List = [
-    [r'defined\s+(\w+)',         '"\\1" in __dict__'],
-    [r'defined\s*\((\w+)\)',     '"\\1" in __dict__'],
-    [r'/\*.*\*/',                ''],
-    [r'/\*.*',                   ''],
-    [r'//.*',                    ''],
-    [r'(0x[0-9A-Fa-f]*)[UL]+',   '\\1'],
+    [r'defined\s+(\w+)',                 '"\\1" in __dict__'],
+    [r'defined\s*\((\w+)\)',             '"\\1" in __dict__'],
```
Comment (bdbaddog) [inline src/engine/SCons/cpp.py:158]: This is changed.

File: src/engine/SCons/cpp.py [hunk 4]
```diff
@@ -234,10 +242,11 @@ def __call__(self, *values):
 
 
 class PreProcessor(object):
+
     """
     The main workhorse class for handling C pre-processing.
     """
-    def __init__(self, current=os.curdir, cpppath=(), dict={}, all=0):
+    def __init__(self, current=os.curdir, cpppath=(), dict={}, all=0, depth=-1):
```
Comment (bdbaddog) [inline src/engine/SCons/cpp.py:249]: This is changed.

Comment (bdbaddog) [reply src/engine/SCons/cpp.py:249]: This is changed.

File: src/engine/SCons/cpp.py [hunk 5]
```diff
@@ -255,9 +264,16 @@ def __init__(self, current=os.curdir, cpppath=(), dict={}, all=0):
...
         self.cpp_namespace['__dict__'] = self.cpp_namespace
 
+        # Return all includes without resolving
         if all:
            self.do_include = self.all_include
 
+        # Max depth of nested includes:
+        # -1 = unlimited
+        # 0 - disabled nesting
+        # >0 - number of allowed nested includes
```
Comment (bdbaddog) [inline src/engine/SCons/cpp.py:274]: This is changed.

**Comment by bdbaddog** [comment]
Can you hop on discord  #devel channel?

**Comment by bdbaddog** [comment]
Sorry for my confusion on this PR.
It is as you said.
Also did some perf testing and no diff.
Merging.

**Comment by ivankravets** [comment]
@bdbaddog Thanks for the merge!!! I'm so proud of this contribution!

> Can you hop on discord #devel channel?

Sorry, it looks like I missed this comment :((((

**Comment by mwichmann** [comment]
Just a little curious... `test/CPPDEFINES/scan.py` emits a line
>Scanner evaluation of CPP lines not yet supported; skipping test.

Do these changes affect that decision in any way?

**Comment by ivankravets** [comment]
This commit should not affest `test/CPPDEFINES/scan.py` because it does not use conditional scanner. Maybe I don't know something :( How to debug this case?

**Comment by mwichmann** [comment]
No, I didn't mean you broke it; it's been this way (intentionally skipped) for a while. I wondered if enough scanning is available now that it could maybe be enabled for this scanner?

**Comment by ivankravets** [comment]
Yes, it makes sense. There are some examples in src/engine/SCons/cppTests.py

We have been using it for ~4 years in production. Maybe our use case will help someone https://github.com/platformio/platformio-core/blob/develop/platformio/builder/tools/piolib.py#L97

In @PlatformIO we thing called "Library Dependency Finder. So, thanks to the SCons and this Conditional Scanner, we actually automatically resolve all dependencies based on `#include` directives and build static libraries on-the-fly. Developers just install libraries from our registry and later just use `#include <include/MyLib.h>`, we do the rest of the work in auto mode.

**Pull request merged**